### DISCUSSION TITLE
Multi-Party Computation for DID

### DISCUSSION THREAD

**Discussion by prdlk** [sonr-io/sonr#337] [Five Head]
**How do we utilize multi-wallet encryption amongst all Motor Nodes connected to a User?**

Read some interesting stuff about [Multi-Party Computation (MPC)](https://www.fireblocks.com/what-is-mpc/). TLDR; MPC enables multiple parties – each holding their own private data – to evaluate a computation without ever revealing any of the private data held by each party (or any otherwise related secret information).
- [MPC based Wallets](https://academy.bit2me.com/en/que-es-wallet-mpc/): With an MPC wallet, the private keys and their control are divided between several devices. This means that in order to use the funds in an MPC wallet, all of the devices must be used. (If you see where I’m getting at)
- [ZenGo Wallet](https://zengo.com/): This is a consumer-facing application for an MPC-based wallet. Right now, it is a two-part system where your devices communicate with ZenGo’s servers

**Solution**
Since root level DID is currently supposed to be using the MultiSig mechanism and it is not implemented yet (along with Race conditions/Edge cases). Therefore, we should do some R&D on how MPC could be integrated with a Motor Node for our transactional mechanism and authorization mechanism. This would allow the Libp2p motor host to operate as the most secure Wallet on the market. Along with this, I was thinking that private key generation should be tied to Users Vault Bucket - tangential to how ZenGo is handling it, except the data is retained with the User (Due to our DID mechanism).

**Resources**
- [Go MPC Library](https://github.com/tuneinsight/lattigo) and [Associated Slideshow](https://homomorphicencryption.org/wp-content/uploads/2020/12/wahc20_slides_christian-1.pdf)
- [Go Multi Party Sig](https://github.com/taurusgroup/multi-party-sig)
- [Awesome MPC](https://github.com/rdragos/awesome-mpc)
- [[WHITEPAPER] MPC Wallet by Unbound](https://github.com/unboundsecurity/blockchain-crypto-mpc/blob/master/docs/Unbound_Cryptocurrency_Wallet_Library_White_Paper.md)
- [[ARTICLE] MPC Primer by ZenGo](https://zengo.com/dive-into-multi-party-computation-mpc-technology-and-its-key-management-in-blockchain-applications/)
- [[VIDEO] MPC Basics](https://www.youtube.com/watch?v=90jcXCHsBF0)

**Diagram(s)**

![Image from iOS](https://user-images.githubusercontent.com/3804558/169713776-09bb102c-58a8-4b84-aa1d-047dc4440ac1.jpg)

**Comment by prdlk** [comment]
https://securecomputation.org/

**Comment by mcjcloud** [comment]
Looking into to MPC more, I don't see how it will be useful to us. Here is what I understand about what MPC does compared with what we need for our use case.

## Our Problem
We need a means of sharing encrypted data between authorized devices, when any number of those devices are offline. We need the ability to "onboard" new devices so that they may also read any of that account's data. So to recap requirements:
* Encrypt data from one device and decrypt on another
* Decrypt data while all other devices are offline
* Add new devices which can also decrypt that data when all devices are offline

## MPC
MPC is a mechanism which allows sensitive information to be withheld, but still used to compute the result of a function on that information. Any invocation of MPC requires that all parties involved be online. It also is not a quick process. So the issues this creates with our use case is:
* Data will not be readable when other devices are offline
* Reading data will be very slow
* Adding new verification methods will be slow and require all other devices to consent

**Reply by prdlk** [reply]
@joshLong145 ?

**Reply by joshLong145** [reply]
Agreed, by utilizing MPC we will be receiving a trade off on computation time for compute of these keys. I have seen this implementation used in distributed ledgers for the use case of transferring of ownership between parties, where time of calculation does not need to be in the milliseconds or shorter. However, I believe we gain a more reliable system by Using MPC for the following reasons: 

* We no longer have to have a trusted authority for key generation (although motor nodes could do this already, we adopt a distributed implementation for our key generation)
* We now can use a derived key pair from all parties (registered devices) to then sign and encrypt objects.

### Main Concerns
- I am unsure how the PK from the VM is now used if the public and private keys will derive from MPC operation.
- If a device is lost or has to be unregistered from a user document, do we then have to regenerate they keys? 
    - decrypt data, then sign and encrypt with new key and store once again? perhaps better way of handling this but this is a concern i have.
- Will the resulting loss in key generation be too much to be considered reasonable (if key generation takes > 1min then we might be reaching a time frame that is too noticeable to users.

**Reply by mcjcloud** [reply]
MPC doesn't really make us any more decentralized than we are. Our whole infrastructure is decentralized, so storing keys in buckets/objects doesn't centralize them. You mentioned we no longer need a trusted authority for key generation, but that's already how it works. Key generation is done by motor, no "authority".

I see the slowness and the fact that every device has to participate in the decryption of data a non-starter. It may be secure but it will have a bad user experience.

Lastly, in all my researching I don't see anything concrete on how MPC will actually create a shared key. If it does actually do this, it seems much more efficient to just pass a shared key between two devices than to go through MPC.

**Reply by joshLong145** [reply]
> MPC doesn't really make us any more decentralized than we are. Our whole infrastructure is decentralized, so storing keys in buckets/objects doesn't centralize them. You mentioned we no longer need a trusted authority for key generation, but that's already how it works. Key generation is done by motor, no "authority".
> 
True, i was more thinking of a highway needing to be involved, but in fact it would not need to be invovled in the computation of the keys, it doesn't have to leave the motor. 
> I see the slowness and the fact that every device has to participate in the decryption of data a non-starter. It may be secure but it will have a bad user experience.

I don't know if all parties will have to participate in decryption as the key will already be created, and stored. which i think means any party can then take advantage
> 
> Lastly, in all my researching I don't see anything concrete on how MPC will actually create a shared key. If it does actually do this, it seems much more efficient to just pass a shared key between two devices than to go through MPC.

https://github.com/taurusgroup/multi-party-sig is an implementation of multi party signing using MPC. this uses ECDSA or Schnorr signatures as key types. this supports both two party and multi party key generation.

**Reply by mcjcloud** [reply]
> I don't know if all parties will have to participate in decryption as the key will already be created, and stored. which i think means any party can then take advantage

How does each device get the generated key? What's the point of having MPC generate a key if it's not involved in the distribution?

**Reply by joshLong145** [reply]
> How does each device get the generated key? What's the point of having MPC generate a key if it's not involved in the distribution?

You are right in that each party member is only given a single portion of its key. which means we would have to form the entire key private key. however, each party is given the public key and verification methods for each point. I could see this method somewhat simplifying our method of caching encrypted jwk's in a map to a single encrypted jwk but that is i think the biggest optimization it would bring. I am now worried that these curve calculations will be rather taxing and be too slow for our cases.

**Reply by mcjcloud** [reply]
We can also just use a single JWK without MPC, since we're going to be using a root level key pair.

**Comment by mcjcloud** [comment]
The one usecase I do see for MPC is for transferring assets like NFTs or aliases on Sonr. This way we can have MFA through multiple devices to authorize those transactions.

**Reply by joshLong145** [reply]
> 

I think i agree that its a better use case but wouldn't this almost always be just two parties involved?

**Comment by prdlk** [comment]
Huzzah @mcjcloud was convinced
